ppb_vector/vector2.py

```diff
@@ -1,97 +1,174 @@
+    """
+    # Basically, see what classes are unique in each type's MRO and return who
+    # has the most.
+    lmro = set(left.__mro__)
+    rmro = set(right.__mro__)
+    lspecial = lmro - rmro
+    rspecial = rmro - lmro
+    if len(lmro) > len(rmro):
+        return left
+    elif len(rmro) > len(lmro):
+        return right
+    else:
+        # They're equal, just arbitrarily pick one
+        return left
+
 
-def _mkvector(value, *, castto=_fakevector):
-    if isinstance(value, Vector2):
-        return value
-    # FIXME: Allow all types of sequences
-    elif isinstance(value, (list, tuple)) and len(value) == 2:
-        return castto(value[0], value[1])
-    # FIXME: Allow all types of mappings
-    elif isinstance(value, dict) and 'x' in value and 'y' in value and len(value) == 2:
-        return castto(value['x'], value['y'])
+def _find_lowest_vector(left: typing.Type, right: typing.Type) -> typing.Type:
+    if left is right:
+        return left
+    elif not issubclass(left, Vector2):
+        return right
+    elif not issubclass(right, Vector2):
+        return left
     else:
-        raise ValueError(f"Cannot use {value} as a vector-like")
+        return _find_lowest_type(left, right)
 
 
-class Vector2(Sequence):
+class Vector2:
+    x: float
+    y: float
+    _length: float
 
-    def __init__(self, x: Real, y: Real):
-        self.x = x
-        self.y = y
-        self.length = hypot(x, y)
+    def __init__(self, x: typing.SupportsFloat, y: typing.SupportsFloat):
+        try:
+            self.x = x.__float__()
+        except AttributeError:
+            raise TypeError(f"{type(x).__name__} object not convertable to float")
+        try:
+            self.y = y.__float__()
+        except AttributeError:
+            raise TypeError(f"{type(y).__name__} object not convertable to float")
 
     @classmethod
-    def convert(cls, value: VectorLike) -> 'Vector2':
+    def convert(cls: typing.Type[VectorOrSub], value: VectorLike) -> VectorOrSub:
         """
         Constructs a vector from a vector-like.
         """
-        return _mkvector(value, castto=type(cls))
+        # Use Vector2.convert() instead of type(self).convert() so that 
+        # _find_lowest_vector() can resolve things well.
+        if isinstance(value, cls):
+            return value
+        elif isinstance(value, Vector2):
+            return cls(value.x, value.y)
+        # FIXME: Allow all types of sequences
+        elif isinstance(value, (list, tuple)) and len(value) == 2:
+            return cls(value[0].__float__(), value[1].__float__())
+        # FIXME: Allow all types of mappings
+        elif isinstance(value, dict) and 'x' in value and 'y' in value and len(value) == 2:
+            return cls(value['x'].__float__(), value['y'].__float__())
+        else:
+            raise ValueError(f"Cannot use {value} as a vector-like")
+
+    @property
+    def length(self) -> float:
+        if not hasattr(self, '_length'):
```

Comment:
Is the reflection dance with `hasattr` to memoize the vector length really faster than recomputing it every time?

Reply:
OK, I did a quick [μbench](https://gist.github.com/nbraud/2dd0ae4904f7e97727f51722a1bba007) with `timeit`, and it looks like the speedup/slowdown of saving the result in `_length` is below the measurement noise, so I would recommend removing that code and just doing `return hypot(self.x, self.y)`.

```
$ ./dunder_bench.py
1 successive length computations
        Memoized: 0.41153700300492346
        Computed: 0.41805144600220956
        Speedup:  1.5829543758445919%
5 successive length computations
        Memoized: 1.2452478090126533
        Computed: 1.2318236450082622
        Speedup:  -1.0780315297269971%
10 successive length computations
        Memoized: 2.252152765984647
        Computed: 2.2965358549845405
        Speedup:  1.970696200996347%
25 successive length computations
        Memoized: 5.274616469978355
        Computed: 5.400348082999699
        Speedup:  2.3837110003537276%
50 successive length computations
        Memoized: 10.367006862012204
        Computed: 10.45787951600505
        Speedup:  0.8765563214376826%
100 successive length computations
        Memoized: 21.056249847984873
        Computed: 21.302848722989438
        Speedup:  1.1711433744606925%
./dunder_bench.py  81.75s user 0.00s system 99% cpu 1:21.77 total
```

Reply:
Does computing it have any significant overhead overhead? I did this to get rid of  `_fakevector` while keeping the performance benefit of not computing the length at construction time.

Reply:
@astronouth7303 That's what the microbenchmark I wrote was showing: even if you access the computed length 100 times in a row, loading it from `self._length` isn't faster than recomputing it.

Reply:
I simplified the implementation of `length`.